internal/pkg/cli/deploy/workload.go

```diff
@@ -367,24 +388,71 @@ func (d *workloadDeployer) uploadContainerImages(out *UploadArtifactsOutput) err
 	if err != nil {
 		return fmt.Errorf("login to image repository: %w", err)
 	}
+
+	var digestsMu sync.Mutex
 	out.ImageDigests = make(map[string]ContainerImageIdentifier, len(buildArgsPerContainer))
+	bufferIdx := 0
+	labeledBuffers := make([]*syncbuffer.LabeledSyncBuffer, len(buildArgsPerContainer))
+	g, ctx := errgroup.WithContext(context.Background())
+	cursor := cursor.New()
+	cursor.Hide()
+
 	for name, buildArgs := range buildArgsPerContainer {
+		// create a copy of loop variables to avoid data race.
+		name := name
+		buildArgs := buildArgs
+
 		buildArgs.URI = uri
 		buildArgsList, err := buildArgs.GenerateDockerBuildArgs(dockerengine.New(exec.NewCmd()))
 		if err != nil {
-			return fmt.Errorf("generate docker build args: %w", err)
-		}
-		// TODO(adi) handle this log in syncBuffer's Print function
-		log.Infof("Building your container image: docker %s\n", strings.Join(buildArgsList, " "))
-		digest, err := d.repository.BuildAndPush(buildArgs)
-		if err != nil {
-			return fmt.Errorf("build and push image: %w", err)
+			return fmt.Errorf("generate docker build args for %q's image: %w", name, err)
 		}
-		out.ImageDigests[name] = ContainerImageIdentifier{
-			Digest:            digest,
-			CustomTag:         d.image.CustomTag,
-			GitShortCommitTag: d.image.GitShortCommitTag,
+		buf := syncbuffer.New()
+		labeledBuffer := buf.WithLabel(fmt.Sprintf("Building your container image %s: docker %s", name, strings.Join(buildArgsList, " ")))
+		labeledBuffers[bufferIdx] = labeledBuffer
+		bufferIdx++
+		pr, pw := io.Pipe()
+		g.Go(func() error {
+			defer pw.Close()
+			digest, err := d.repository.BuildAndPush(ctx, buildArgs, pw)
+			if err != nil {
+				return fmt.Errorf("build and push the image for %q container: %w", name, err)
+			}
+			digestsMu.Lock()
```

Comment:
```suggestion
			digestsMu.Lock()
			defer digestsMu.Unlock()
```

Reply:
i thought it is good to unlock the map before `return nil`